src/main/java/org/candlepin/pki/certs/SCACertificateGenerator.java

```diff
@@ -58,216 +65,333 @@
-        existing.setCert(this.pemEncoder.encodeAsString(x509Cert));
-        existing.setContent(this.createPayloadAndSignature(payloadBytes));
+            KeyPair keypair = keyPairGenerator.getKeyPair(consumer);
 
-        SCACertificate savedCert = this.contentAccessCertificateCurator.create(existing);
-        consumer.setContentAccessCert(savedCert);
-        this.consumerCurator.merge(consumer);
-        return savedCert;
-    }
+            scaCertificate = new SCACertificate();
+            scaCertificate.setConsumer(consumer);
+            scaCertificate.setKeyAsBytes(pemEncoder.encodeAsBytes(keypair.getPrivate()));
+            scaCertificate.setCert(""); // we'll correct this later
+
+            scaCertificate = contentAccessCertificateCurator.create(scaCertificate);
+
+            consumer.setContentAccessCert(scaCertificate);
+        }
 
-    private SCACertificate updateCertificate(Consumer consumer,
-        Owner owner, SCACertificate existing) {
-        Date now = new Date();
-        Date expiration = existing.getSerial().getExpiration();
-        boolean isX509CertExpired = expiration.before(now);
+        Date regenerationCutoff = Date.from(Instant.now()
+            .plus(Long.valueOf(x509CertExpirationThreshold), ChronoUnit.DAYS));
+
+        CertificateSerial serial = scaCertificate.getSerial();
+        if (serial == null ||
+            serial.getExpiration() == null ||
+            regenerationCutoff.after(serial.getExpiration())) {
+
+            log.info("Generating new SCA x509 certificate for consumer: \"{}\"", consumer.getUuid());
+            if (serial != null) {
+                serialCurator.revokeById(serial.getId());
+            }
 
-        if (isX509CertExpired) {
             OffsetDateTime start = OffsetDateTime.now().minusHours(1L);
             OffsetDateTime end = start.plusYears(1L);
 
-            KeyPair keyPair = this.keyPairGenerator.getKeyPair(consumer);
-            this.serialCurator.revokeById(existing.getSerial().getId());
-            CertificateSerial serial = createSerial(end);
-            org.candlepin.model.dto.Product container = createSCAProdContainer(owner, consumer);
-            X509Certificate x509Cert = createX509Cert(consumer.getUuid(), owner,
-                serial, keyPair, container, start, end);
+            serial = createSerial(end);
+            KeyPair keypair = keyPairGenerator.getKeyPair(consumer);
+            org.candlepin.model.dto.Product container = createProductContainer(owner, environments);
+
+            X509Certificate x509Cert = createX509Cert(consumer.getUuid(), owner, serial, keypair,
+                container, start, end);
 
-            existing.setSerial(serial);
-            existing.setCert(this.pemEncoder.encodeAsString(x509Cert));
-            this.contentAccessCertificateCurator.saveOrUpdate(existing);
+            scaCertificate.setSerial(serial);
+            scaCertificate.setCert(pemEncoder.encodeAsString(x509Cert));
```

Comment:
we did a `curator.create()` earlier; don't we need to also do a merge after setting this data?

Reply:
Nope. So long as the entity is managed and a new instance hasn't been spun up or anything, we need not do another merge. The concern would be if we ever shut off the auto commit stuff, then we probably would need it.